### ISSUE TITLE
flex_attention can spend >10x time in create_block_mask vs actual attention for document mask settings?

### ISSUE THREAD

**Issue by Lunarmony** [pytorch-labs/attention-gym#149]
Hi! 

I noticed `flex_attention` can spend significantly more time on creating BlockMask vs doing the actual attention operations sometimes. I have a minimal repro code below where flex_attention spent 5271ms on creating block mask vs fwd (412 ms). This does not happen if I rewrite without document masks / jagged tensors, and the gap can increase even more if I make `mask_mod` more complex. Would you happen to have suggestions on how to improve the efficiency?

Environment: pytorch 2.7.1, cuda 12.8, behavior reproducible on both 6000 Adas and H200s.

Benchmark code:

```python
import time

import torch
from pyre_extensions import none_throws
from torch.nn.attention.flex_attention import (
    BlockMask,
    create_block_mask,
    flex_attention,
)

torch.set_default_device("cuda")
torch._dynamo.config.capture_scalar_outputs = True
create_block_mask = torch.compile(create_block_mask, dynamic=True)
flex_attention_compiled = torch.compile(flex_attention, dynamic=True)


def lengths_to_offsets(lengths: torch.Tensor) -> torch.Tensor:
    offsets = torch.zeros(
        lengths.shape[0] + 1, dtype=torch.int32, device=lengths.device
    )
    torch.cumsum(lengths, dim=0, out=offsets[1:])
    return offsets


def create_block_mask_for_flex_attention(
    offsets: torch.Tensor,
) -> BlockMask:
    device = offsets.device
    batch_size = len(offsets) - 1
    torch._check_is_size(batch_size)
    torch._check(batch_size >= 1)

    # Generate local document IDs for jagged tensors
    combined_seq_len: int = offsets[-1].item()  # \sum b L_b
    # print(f"combined_seq_len: {combined_seq_len}, dtype: {type(combined_seq_len)}")
    torch._check_is_size(combined_seq_len)
    torch._check(combined_seq_len >= 1)

    lengths = offsets[1:] - offsets[:-1]  # (B,)
    doc_ids = torch.repeat_interleave(
        torch.arange(batch_size, dtype=torch.int64, device=device),
        lengths,
    )  # (\sum b L_b,)

    def _causal(
        b: torch.Tensor, h: torch.Tensor, q_idx: torch.Tensor, kv_idx: torch.Tensor
    ) -> torch.Tensor:
        same_doc = doc_ids[q_idx] == doc_ids[kv_idx]
        causal = q_idx >= kv_idx
        return same_doc & causal

    return create_block_mask(
        mask_mod=_causal,
        B=None,
        H=None,
        Q_LEN=combined_seq_len,
        KV_LEN=combined_seq_len,
    )


class SimpleAttention(torch.nn.Module):
    """Minimal implementation of attention with FlexAttention for jagged tensors."""

    def __init__(
        self,
        dim: int,
        num_heads: int,
        num_kv_heads: int,
    ):
        super().__init__()

        self._dim = dim
        self._num_heads = num_heads
        self._num_kv_heads = num_kv_heads if num_kv_heads is not None else num_heads
        self._head_dim = dim // num_heads
        self._qkv = torch.nn.Linear(
            dim, (self._num_heads + 2 * self._num_kv_heads) * self._head_dim, bias=False
        )
        self._out_proj = torch.nn.Linear(dim, dim, bias=False)

    def forward(
        self,
        x: torch.Tensor,  # (\\sum b L_b, D)
        x_offsets: torch.Tensor,  # (B+1,)
        block_mask: BlockMask,
    ) -> torch.Tensor:
        """Forward pass for jagged tensors."""
        combined_seq_len = x.shape[0]
        qkv = self._qkv(x)
        q_dim = self._num_heads * self._head_dim
        k_dim = v_dim = self._num_kv_heads * self._head_dim
        q, k, v = torch.split(qkv, [q_dim, k_dim, v_dim], dim=-1)

        # Reshape for attention
        q = q.reshape(combined_seq_len, self._num_heads, self._head_dim).permute(
            1, 0, 2
        )  # (H, N, D)
        k = k.reshape(combined_seq_len, self._num_kv_heads, self._head_dim).permute(
            1, 0, 2
        )  # (H, N, D)
        v = v.reshape(combined_seq_len, self._num_kv_heads, self._head_dim).permute(
            1, 0, 2
        )  # (H, N, D)

        q = q.unsqueeze(0)  # (1, H, N, D)
        k = k.unsqueeze(0)  # (1, H, N, D)
        v = v.unsqueeze(0)  # (1, H, N, D)

        attn_output = flex_attention_compiled(
            q,
            k,
            v,
            block_mask=block_mask,
            enable_gqa=True,
        )
        attn_output = attn_output.squeeze(0)  # (H, N, D)
        attn_output = attn_output.permute(1, 0, 2).reshape(
            combined_seq_len, -1
        )  # (N, H*D)
        return self._out_proj(attn_output)


def padded_to_ragged(
    values: torch.Tensor, lengths: torch.Tensor
) -> tuple[torch.Tensor, torch.Tensor]:
    """Convert a padded tensor to jagged format."""
    batch_size, max_len, dim = values.shape
    device = values.device

    offsets = lengths_to_offsets(lengths)

    jagged_values = torch.zeros((offsets[-1], dim), dtype=values.dtype, device=device)

    for i in range(batch_size):
        start = offsets[i]
        end = offsets[i + 1]
        seq_len = lengths[i]
        jagged_values[start:end] = values[i, :seq_len]

    return jagged_values, offsets


def run_config(
    batch_size: int,
    max_seq_len: int,
    dim: int,
    num_heads: int,
    num_kv_heads: int,
    num_microbatches: int = 1,
    config_name: str = "",
) -> torch.Tensor:
    torch.manual_seed(42)
    microbatches_x = []
    microbatches_lengths = []
    microbatches_x_values = []
    microbatches_x_offsets = []

    for i in range(num_microbatches):
        x = torch.randn(batch_size, max_seq_len, dim, device="cuda")
        min_length = max_seq_len // 2
        lengths = torch.randint(
            min_length, max_seq_len + 1, (batch_size,), device="cuda", dtype=torch.int32
        )
        microbatches_x.append(x)
        microbatches_lengths.append(lengths)
        print(
            f"Microbatch {i} sequence lengths: {lengths} / total tokens: {microbatches_lengths[i].sum().item()}"
        )
        x_values, x_offsets = padded_to_ragged(
            microbatches_x[i], microbatches_lengths[i]
        )
        microbatches_x_values.append(x_values)
        microbatches_x_offsets.append(x_offsets)

    # Create attention model
    model = SimpleAttention(
        dim=dim,
        num_heads=num_heads,
        num_kv_heads=num_kv_heads,
    ).to("cuda")

    # Warm-up compile to exclude one-time compilation overhead
    _block_mask = create_block_mask_for_flex_attention(x_offsets)
    _ = model(x_values, x_offsets, _block_mask)
    torch.cuda.synchronize()

    # Timing measurements across microbatches
    block_mask_times: list[float] = []
    attention_times: list[float] = []
    output = None  # Will hold last microbatch output

    for x_values, x_offsets in zip(microbatches_x_values, microbatches_x_offsets):
        # Measure block mask creation
        torch.cuda.synchronize()
        start = time.perf_counter()
        block_mask = create_block_mask_for_flex_attention(x_offsets)
        torch.cuda.synchronize()
        block_mask_times.append(time.perf_counter() - start)

        # Measure attention forward
        torch.cuda.synchronize()
        start = time.perf_counter()
        output = model(x_values, x_offsets, block_mask)
        torch.cuda.synchronize()
        attention_times.append(time.perf_counter() - start)

    avg_block_mask_ms = 1000.0 * sum(block_mask_times) / num_microbatches
    avg_attention_ms = 1000.0 * sum(attention_times) / num_microbatches

    print(
        f"Average block mask creation time over {num_microbatches} microbatches: "
        f"{avg_block_mask_ms:.3f} ms"
    )
    print(
        f"Average attention forward time over {num_microbatches} microbatches: "
        f"{avg_attention_ms:.3f} ms"
    )

    print(f"Output shape: {none_throws(output).shape}")
    print(f"Test with {config_name} passed successfully.")
    return output


if __name__ == "__main__":
    run_config(
        batch_size=4,
        max_seq_len=512,
        dim=128,
        num_heads=4,
        num_kv_heads=2,
        num_microbatches=5,
        config_name="small sequence length (512x4)",
    )

    try:
        run_config(
            batch_size=32,
            max_seq_len=32768,
            dim=128,
            num_heads=4,
            num_kv_heads=2,
            num_microbatches=5,
            config_name="large sequence length (32768x32)",
        )
    except Exception as e:
        pass
```

Outputs:

```bash
python flex_attention_repro.py 2>&1 | tee ~/flex_attention_repro.log
...
Microbatch 0 sequence lengths: tensor([27193, 17105, 27738, 32728, 29049, 31005, 18471, 24557, 25752, 27961,
        25953, 24787, 27006, 27607, 20068, 29350, 20461, 26167, 24311, 28586,
        31781, 30480, 25548, 18091, 23404, 17366, 16850, 25592, 28147, 21291,
        30769, 21488], device='cuda:0', dtype=torch.int32) / total tokens: 806662
Microbatch 1 sequence lengths: tensor([32005, 17848, 22068, 31229, 22934, 22262, 17059, 16930, 31708, 25358,
        17511, 18766, 28742, 17742, 26016, 17028, 28834, 24833, 24229, 19796,
        25687, 24142, 26345, 22302, 19873, 28462, 25351, 28298, 17052, 18595,
        32313, 31383], device='cuda:0', dtype=torch.int32) / total tokens: 762701
Microbatch 2 sequence lengths: tensor([24035, 32147, 22040, 22195, 20340, 27302, 28200, 32236, 31099, 18167,
        19978, 24703, 27661, 27089, 22584, 20082, 20823, 27175, 26150, 23806,
        31934, 19483, 23370, 19470, 16762, 16950, 26906, 25280, 16508, 26315,
        18385, 25223], device='cuda:0', dtype=torch.int32) / total tokens: 764398
Microbatch 3 sequence lengths: tensor([20112, 31489, 25993, 23811, 26373, 27193, 29394, 21839, 26525, 29203,
        30526, 18685, 29974, 23247, 21164, 25957, 24852, 18251, 17853, 18487,
        28513, 30936, 18299, 25019, 20684, 22231, 16450, 27899, 27967, 21378,
        19516, 24205], device='cuda:0', dtype=torch.int32) / total tokens: 774025
Microbatch 4 sequence lengths: tensor([31067, 24878, 24461, 19486, 19374, 30672, 30496, 24621, 25211, 17570,
        18779, 25981, 30860, 29057, 27960, 26127, 20553, 22116, 27735, 20797,
        29972, 17528, 17804, 21020, 27538, 25174, 29886, 19396, 20474, 26769,
        28902, 20231], device='cuda:0', dtype=torch.int32) / total tokens: 782495
Average block mask creation time over 5 microbatches: 5271.150 ms
Average attention forward time over 5 microbatches: 412.081 ms
```

**Comment by drisspg** [comment]
Case 1 is expected, as well I have added a script that you can use to:
https://github.com/pytorch-labs/attention-gym/blob/main/benchmarks/bench_block_mask.py

The problem with 1 is that by default we will have a data structure that is O(N**2) in memory and we use compile to not need to produce this.


Case 2:
That is a little surprising can you get a full profile to try and pin point if perhaps there is some extra cpu overhad that is creeping in?

**Comment by drisspg** [comment]
@Lunarmony I just added an example of how to do this for causal attention: https://github.com/pytorch-labs/attention-gym/pull/155

This `from_kv_blocks` will create the q_num_blocks that is needed for backwards. If you are only focused on inference you can use:

https://github.com/pytorch/pytorch/blob/7e433d5f423248914c5e9838d3ea145db7964923/torch/nn/attention/flex_attention.py#L352

which just landed to avoid the computation of them
